### ISSUE TITLE
no-misused-generics: false positive?

### ISSUE THREAD

**Issue by AlCalzone** [fimbullinter/wotan#520]
I have the following pattern to provide correctly typed event signatures for a class that extends `EventEmitter`:

```ts
export interface ConnectionWatcher {
	on<TEvent extends ConnectionWatcherEvents, TCb = ConnectionWatcherEventCallbacks[TEvent]>(event: TEvent, callback: TCb): this;
	removeListener<TEvent extends ConnectionWatcherEvents, TCb = ConnectionWatcherEventCallbacks[TEvent]>(event: TEvent, callback: TCb): this;
	removeAllListeners(event?: ConnectionWatcherEvents): this;
}

export type ConnectionWatcherEvents =
	"ping succeeded" | "ping failed" |
	"connection alive" | "connection lost" |
	"gateway offline" |
	"reconnecting" |
	"give up"
	;

export type PingFailedCallback = (failedPingCount: number) => void;
export type ReconnectingCallback = (reconnectAttempt: number, maximumReconnects: number) => void;

export type ConnectionWatcherEventCallbacks = Merge<
	{ [K in ConnectionWatcherEvents]: () => void },
	{
		"ping failed": PingFailedCallback;
		"reconnecting": ReconnectingCallback;
	}
	>;
```
Basically the events `"ping failed"` and `"reconnecting"` have a special signature while the others are just `() => void`.

`wotan` flags all generic parameters of `ConnectionWatcher` with `TypeParameter 'TEvent' is not used to enforce a constraint between types and can be replaced with 'ConnectionWatcherEvents'`. However I believe the constraints are necessary to achieve this behavior:
![unbenannt](https://user-images.githubusercontent.com/17641229/50897616-f25f3680-140c-11e9-81ac-51d56c8ad48e.PNG)

**Comment by ajafff** [comment]
Note that there's no **constraint** on `TCb`. So this error seems correct.
You either want to use `TEvent` in the constraint of `TCb` (in addition to the default), or simply inline `TCb`:

```ts
export interface ConnectionWatcher {
	on<TEvent extends ConnectionWatcherEvents, TCb extends ConnectionWatcherEventCallbacks[TEvent] = ConnectionWatcherEventCallbacks[TEvent]>(event: TEvent, callback: TCb): this;
}

// or
export interface ConnectionWatcher {
	on<TEvent extends ConnectionWatcherEvents>(event: TEvent, callback: ConnectionWatcherEventCallbacks[TEvent]): this;
}
```

**Comment by AlCalzone** [comment]
I see... I've tried inlining both, but that just gave me a union of callbacks. Inlining `TCb` works :+1:
